r2-core/src/main/java/com/linkedin/r2/filter/transport/ServerRetryFilter.java

```diff
@@ -75,13 +116,160 @@ public void onStreamError(Throwable ex,
+        if (_serverRetryTracker.isBelowRetryRatio())
+        {
+          LOG.debug("RetriableRequestException caught! Do retry. Error message: {}", message);
+          wireAttrs.put(R2Constants.RETRY_MESSAGE_ATTRIBUTE_KEY, message);
+        }
+        else
+        {
+          LOG.debug("Max request retry ratio exceeded! Will not retry. Error message: {}", message);
+          wireAttrs.remove(R2Constants.RETRY_MESSAGE_ATTRIBUTE_KEY);
+        }
         break;
       }
       cause = cause.getCause();
     }
 
     nextFilter.onError(ex, requestContext, wireAttrs);
   }
+
+  private void updateRetryTracker(Request req)
+  {
+    String retryAttemptsHeader = req.getHeader(HttpConstants.HEADER_NUMBER_OF_RETRY_ATTEMPTS);
+    if (retryAttemptsHeader != null)
+    {
+      _serverRetryTracker.add(Integer.parseInt(retryAttemptsHeader));
+    }
+  }
+
+  /**
+   * Stores the number of requests categorized by number of retry attempts. It uses the information to estimate
+   * a ratio of how many requests are being retried in the cluster. The ratio is then compared with
+   * {@link ServerRetryTracker#_maxRequestRetryRatio} to make a decision on whether or not to retry in the
+   * next interval. When calculating the ratio, it looks at the last {@link ServerRetryTracker#_aggregatedIntervalNum}
+   * intervals by aggregating the recorded requests.
+   */
+  private static class ServerRetryTracker
+  {
+    private final int _retryLimit;
+    private final int _aggregatedIntervalNum;
+    private final double _maxRequestRetryRatio;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private boolean _isBelowRetryRatio;
+
+    @GuardedBy("_counterLock")
+    private final LinkedList<int[]> _retryAttemptsCounter;
+    private final int[] _aggregatedRetryAttemptsCounter;
+
+    private ServerRetryTracker(int retryLimit, int aggregatedIntervalNum, double maxRequestRetryRatio, long updateIntervalMs, Clock clock)
+    {
+      _retryLimit = retryLimit;
+      _aggregatedIntervalNum = aggregatedIntervalNum;
+      _maxRequestRetryRatio = maxRequestRetryRatio;
+      _updateIntervalMs = updateIntervalMs;
+      _clock = clock;
+
+      _aggregatedRetryAttemptsCounter = new int[_retryLimit + 1];
+      _retryAttemptsCounter = new LinkedList<>();
+      _retryAttemptsCounter.add(new int[_retryLimit + 1]);
+    }
+
+    public void add(int numberOfRetryAttempts)
+    {
+      if (numberOfRetryAttempts <= _retryLimit)
+      {
+        synchronized (_counterLock)
+        {
+          _retryAttemptsCounter.getLast()[numberOfRetryAttempts] += 1;
+        }
+      } else
+      {
+        LOG.warn("Unexpected number of retry attempts: " + numberOfRetryAttempts + ", current retry limit: " + _retryLimit);
```

Comment:
Is this protecting against malicious clients? Why would we be getting a value that is above the limit? If something is misconfigured and clients are trying more than they should I would think we would want to track this as another entry for the max number of retries that are allowed instead of ignoring it.

Reply:
This is to account for a config discrepancy between client and server. Client and server can both configure their own retry limit, so potentially, the client-side config can have a higher value than the server-side config. Do we want to fail the request if the number of attempts is higher than the server-side config value? Or should the server still handle this request? @rachelhanhan @bbarkley @ssheng Any thoughts on this?

I think not adding this request to RetryTracker will help improve efficiency because we can use a fix-sized array instead of a resizable list. And really, requests with high retry attempts should be extremely rare. We can already infer the retry ratio from the requests with lower retry attempts.

Reply:
I agree we shouldn't support tracking data for any number of retry requests - if the current count is greater than the retry limit we can just treat it as the retry limit max.
Though potentially we should be rejecting the request if the count is higher than what the service wants to allow. Is the server side limit used for anything other than tracking?

Reply:
Fixed by treating it as the retry limit max. The server side limit is used only for tracking now. One reason for the server to still handle the request above the limit is that it already defines through TransportClientProperties the max ratio of retry request a client could send. I think it's reasonable to let the client decide the per-request retry limit.